### PR TITLE
Avoid unnecessary object deep copy by using read-only get/list functions in controllers

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/6866

### CODE REVIEW

**Comment by innobead** [comment]
@shuo-wu PR check failed, please resolve that.

**Comment by innobead** [comment]
@shuo-wu is this ready to review?

**Comment by shuo-wu** [comment]
Yes it's ready to review. But I need to do some tests before the merge.

**Comment by innobead** [comment]
> Yes it's ready to review. But I need to do some tests before the merge.

Sounds good. See if we have a way to see the improvement.

File: controller/volume_controller.go [hunk 1]
```diff
@@ -2600,12 +2604,12 @@ func (c *VolumeController) getNodeCandidatesForAutoBalanceZone(v *longhorn.Volum
 		return candidateNames
 	}
 
-	readyNodes, err := c.ds.ListReadyAndSchedulableNodes()
+	readyNodes, err := c.ds.ListReadyAndSchedulableNodesRO()
```
Comment (derekbit) [inline controller/volume_controller.go:2607]: What's the difference between `ListReadyAndSchedulableNodesRO` and `ListReadyAndSchedulableNodes` for `delete(readyNodes, nName)`?

Comment (shuo-wu) [reply controller/volume_controller.go:2607]: Here, the operations like `delete(readyNodes, nName)` do not include object updating over `readyNodes`. Hence, there is no need to do deep copy during the node list

Comment (derekbit) [reply controller/volume_controller.go:2607]: Got it. Thank you.

**Review by derekbit — DISMISSED** [review]

**Comment by shuo-wu** [comment]
Core tests passed
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5200/

**Comment by innobead** [comment]
This will be only introduced to master and 1.5.4+.

**Review by innobead — APPROVED** [review]
In general, LGTM.

**Comment by innobead** [comment]
@mergify backport v1.5.x

**Pull request merged**